bgpd, zebra: Creating Loopback Interface Flaps BGPd, it should update…

kssoman opened the pull request:
… the Router ID

in Next Restart only. #2865

When ZEBRA_ROUTER_ID_UPDATE is received from RIB, check if the current router ID
is valid (the IP address exists in bgp address hash table). If valid do not
update bgp router ID. This will prevent flapping of bgp session due to change
in router ID

Signed-off-by: kssoman <somanks@vmware.com>

kssoman commented:
RCA
------

When IP address is added to or deleted from interfaces in the system, RIB calculates new router ID for clients  (protocols) and sends ZEBRA_ROUTER_ID_UPDATE  to clients.

The router ID is stored in a sorted list for loopback and other interfaces in struct zebra_vrf

struct list *rid_all_sorted_list;
struct list *rid_lo_sorted_list;

The new router ID is calculated as follows :  
          router_id_get(struct prefix *p, vrf_id_t vrf_id)

If global router ID is configured using "router-id <addr> command, it is preferred 
else if loopback interfaces are configured check the loopback router ID list and select highest loopback 
      interface address
else check all interfaces sorted list and select highest  IP address as router ID

Send the selected router ID to clients

BGP gets ZEBRA_ROUTER_ID_UPDATE  and calls bgp_router_id_update()

This calls bgp_router_id_zebra_bump() to store newly received router ID in router_id_zebra in the global BGP structure 

If BGP router ID is not configured statically using "bgp router-id <addr>" command  then call bgp_router_id_set() to update BGP router ID and send NOTIFY message with reason PEER_DOWN_RID_CHANGE 
This results is session flap

Description of the code changes 
-------------------------------------  

When interfaces are added or deleted,  RIB sends ZEBRA_INTERFACE_ADDRESS_ADD or ZEBRA_INTERFACE_ADDRESS_DELETE to clients.

BGP calls the following functions to process the msg : 

ADD : bgp_interface_address_add() -> bgp_connected_add() -> bgp_address_add()  
      
         The address is added to hash table

DELETE : bgp_interface_address_delete() -> bgp_connected_delete() -> bgp_address_del()

            The address is deleted from the hash table

The code changes checks if the current bgp router ID is present in the address hash table when ZEBRA_ROUTER_ID_UPDATE is received
If the address is present then do not update the bgp router ID and flap the session

If address is not present in the hash table, update router ID and flap the session

Also code change is done to send ZEBRA_INTERFACE_ADDRESS_ADD or ZEBRA_INTERFACE_ADDRESS_DELETE before sending ZEBRA_ROUTER_ID_UPDATE
to client which will enable the client to update the hash table before processing the router ID update message

kssoman commented:
Test logs

[rtr_logs.txt](https://github.com/kssoman/frr/files/2309905/rtr_logs.txt)

manoj-viswanath commented on bgpd/bgpd.c:
```diff
@@ -278,17 +279,31 @@ void bgp_router_id_zebra_bump(vrf_id_t vrf_id, const struct prefix *router_id)
-				bgp_router_id_set(bgp, &router_id->u.prefix4);
+			bgp->router_id_zebra = *prefix;
+			if (!bgp->router_id_static.s_addr) {
+				if(bgp_router_id_valid(bgp) < 0) {
```
As i understand, Router ID update from Zebra is not handled if existing router_id is present in BGP Address hash (probably as an interface address).

Let us take a configuration in this sequence:-
------------------------------
router-id 1.1.1.1
Int lo
  ip address 1.1.1.1/32

router-id 2.2.2.2 ----> Zebra updates its router-id and intimates BGP. BGP ignores this as 1.1.1.1 exist in address hash

int lo
  no ip address 1.1.1.1/32 ---> Zebra deletes this from IFC and sends address delete to BGP. BGP does not update its router id still. 
------------------------------
Looks like the handling is not consistent with your fix as the necessary and sufficient condition you had checked is no longer true. DO we need to update the router-id in this case in BGP ?

manoj-viswanath commented on bgpd/bgp_nexthop.c:
```diff
@@ -251,6 +251,22 @@ static void bgp_address_del(struct bgp *bgp, struct prefix *p)
 }
 
+/* Verify if the bgp router ID is in the address hash table */
+int bgp_router_id_valid(struct bgp *bgp)
```
nitpick: Suggest using bool for such true or false APIs

manoj-viswanath commented on bgpd/bgpd.c:
```diff
@@ -270,6 +270,7 @@ void bgp_router_id_zebra_bump(vrf_id_t vrf_id, const struct prefix *router_id)
 {
 	struct listnode *node, *nnode;
 	struct bgp *bgp;
+	struct in_addr *prefix = (struct in_addr *)&router_id->u.prefix4;
```
nitpick: suggest changing name as this is not a "prefix"

manoj-viswanath reviewed (changes requested):
Have suggested a couple of minor comments as well as a query on handling IF Address delete

kssoman commented:
When interface address delete is received, call bgp_router_id_zebra_bump() to update router ID

srimohans commented on bgpd/bgpd.c:
```diff
@@ -278,17 +282,38 @@ void bgp_router_id_zebra_bump(vrf_id_t vrf_id, const struct prefix *router_id)
-
-			if (!bgp->router_id_static.s_addr)
-				bgp_router_id_set(bgp, &router_id->u.prefix4);
+			if(addr)
```
1) Exact same code in both if & else loops. Consider moving this to a function and then call the function from both if & else loops.
2) Can you start adding comments to some of the important functionality ?

srimohans commented on bgpd/bgp_zebra.c:
```diff
@@ -396,12 +398,23 @@ static int bgp_interface_address_delete(int command, struct zclient *zclient,
 			   ifc->ifp->name, buf);
 	}
 
+	/* Update the router id if the interface address being deleted is
```
Is this function called for IP address delete of both normal interface & Loopback interfaces ?

srimohans commented on bgpd/bgpd.c:
```diff
@@ -278,17 +282,38 @@ void bgp_router_id_zebra_bump(vrf_id_t vrf_id, const struct prefix *router_id)
+				addr = &bgp->router_id_zebra;
+
+			if (!bgp->router_id_static.s_addr) {
+				if(bgp_router_id_valid(bgp) == false) {
```
When the execution comes into this function as part of IP address delete (which was configured as router-id), then bgp_router_id_valid(bgp) should always return false and never returns true.  This is because as part of IP address delete, the address will be removed from address_hash and so this function returns false. This is only for delete case though. So can we modify the code accordingly ?

srimohans commented on zebra/redistribute.c:
```diff
@@ -478,14 +477,13 @@ void zebra_interface_address_delete_update(struct interface *ifp,
 
 	zebra_vxlan_add_del_gw_macip(ifp, ifc->address, 0);
 
-	router_id_del_address(ifc);
```
This ordering change may cause some issues. Did you try to test any race condition scenarios for this ?

srimohans reviewed (changes requested):
Please check the comments.

kssoman commented:
Test logs attached

[rtr_logs.txt](https://github.com/kssoman/frr/files/2323560/rtr_logs.txt)

srimohans commented on bgpd/bgpd.c:
```diff
@@ -278,17 +282,27 @@ void bgp_router_id_zebra_bump(vrf_id_t vrf_id, const struct prefix *router_id)
 
-			bgp->router_id_zebra = router_id->u.prefix4;
+			/* If router_id is NULL, then it is interface delete case
+			 * Set bgp router ID to the value router_id_zebra
```
The 2nd comment I mentioned is to move the below lines of code to a common function as this is exactly same.

// ***************/
if(addr)
    bgp->router_id_zebra = *addr;
else
    addr = &bgp->router_id_zebra;
 
if (!bgp->router_id_static.s_addr)
	bgp_router_id_set(bgp, &router_id->u.prefix4);					

/*********************/

But thats fine. I think you can go ahead with the check in process

srimohans reviewed (approved):
Please check the comments and go ahead

Commit:
bgpd, zebra: Creating Loopback Interface Flaps BGPd, it should update
             the Router ID in Next Restart only (#2865)

* Added function bgp_validate_set_router_id() which will check if
  the current bgp router ID is valid (the IP address exists in
  bgp address hash table). If valid do not update bgp router ID.
  This will prevent flapping of bgp session due to change in router ID

* Changed the sequence of sending interface address ADD and DELETE
  and router ID update message. This will help bgp update address
  hash table which is used for router ID validation

Signed-off-by: kssoman <somanks@vmware.com>
